File: objective-c/Main/03_bob/Bob.m
#import "Bob.h"

@implementation Bob

#pragma mark - Constants

static NSString *beThatWay = @"Fine. Be that way!";
static NSString *chillOut = @"Whoa, chill out!";
static NSString *sure = @"Sure.";
static NSString *whatever = @"Whatever.";

#pragma mark - Logic

- (NSString *) hey:(NSString *)input {
    BOOL isOnlyWhitespace = [input stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]].length == 0;
    BOOL isUppercase = [self isUppercase:input];
    BOOL hasQuestionMark = [input hasSuffix:@"?"];

    if (isOnlyWhitespace) {
        return beThatWay;
    }

    if (isUppercase) {
        return chillOut;
    }

    if (hasQuestionMark) {
        return sure;
    }

    return whatever;
}

- (BOOL) isUppercase:(NSString *)text {
    NSString *filtered = [[text componentsSeparatedByCharactersInSet:[[NSCharacterSet letterCharacterSet] invertedSet]] componentsJoinedByString:@""];
    if ([filtered isEqual:@""]) {
        return NO;
    }

    return [filtered isEqualToString:[filtered uppercaseString]];
}

@end

File: swift/Main/09_bob/Sources/Bob/Bob.swift
import Foundation

struct Bob {

    // MARK: - Constants

    private static let beThatWay = "Fine. Be that way!"
    private static let sure = "Sure."
    private static let chillout = "Whoa, chill out!"
    private static let whatever = "Whatever."

    // MARK: - Logic

    static func hey(_ input: String) -> String {
        let isOnlyWhitespace = input.trimmingCharacters(in: .whitespacesAndNewlines).count == 0
        if isOnlyWhitespace {
            return beThatWay
        }

        let isUppercase = self.isUppercase(input)
        if isUppercase {
            return chillout
        }

        let hasQuestionmark = input.hasSuffix("?")
        if hasQuestionmark {
            return sure
        }

        return whatever
    }

    private static func isUppercase(_ input: String) -> Bool {
        let filtered = input.components(separatedBy: CharacterSet.letters.inverted).joined()
        if filtered == "" {
            return false
        }

        return filtered == filtered.uppercased()
    }
}
